### DISCUSSION TITLE
Problems with remote htcondor workflow

### DISCUSSION THREAD

**Discussion by monirehyamrali** [columnflow/columnflow#307] [Q&A] [answered]
Dear all,

I'm new to columnflow and try to produce columns with the UniteColumns task so I can do some tests outside of cf. I want to consider multiple large datasets, which is why I'd like to use htcondor for parallel computations. This is my command

```
law run cf.UniteColumnsWrapper --version v2 --datasets hh_ggf_bbtautau_madgraph,tt_dl_powheg,tt_sl_powheg,tt_fh_powheg,'graviton_hh_ggf_bbtautau_m*_madgraph' --config run2_2017_nano_uhh_v11 --workers 4 --cf.CalibrateEvents-workflow htcondor --cf.SelectEvents-workflow htcondor --cf.MergeReducedEvents-workflow htcondor --cf.ProduceColumns-workflow htcondor
```

The software bundling works fine and submitting the jobs also works as expected. However, I see an issue in the CalibrateEvents htcondor jobs. Here is an example of the output I see

```
job: 14, branches: [13], id: 16195020.0, status: retry, code: 50, error: undefined, job script error: task dependency check failed, log: /nfs/dust/cms/user/yamralim/cf_cache/hbt_store/analysis_hbt/cf.CalibrateEvents/run2_2017_nano_uhh_v11/tt_fh_powheg/nominal/calib__default/v2/stdall_13To14.txt
```

Here is the part of the indicated log file that I think might be relevant for this discussion:

```
-- dependency tree
Traceback (most recent call last):
  File "/var/lib/condor/execute/dir_56729/job_ZhF6p7krUF4M/repo/modules/columnflow/bin/law", line 10, in <module>
    exit_code = run()
  File "/var/lib/condor/execute/dir_56729/job_ZhF6p7krUF4M/repo/modules/columnflow/modules/law/law/cli/cli.py", line 74, in run
    return mods[prog].execute(*exec_args)
  File "/var/lib/condor/execute/dir_56729/job_ZhF6p7krUF4M/repo/modules/columnflow/modules/law/law/cli/run.py", line 57, in execute
    mod = __import__(modid, globals(), locals(), [cls_name])
  File "/var/lib/condor/execute/dir_56729/job_ZhF6p7krUF4M/repo/modules/columnflow/columnflow/__init__.py", line 55, in <module>
    from columnflow.util import maybe_import
  File "/var/lib/condor/execute/dir_56729/job_ZhF6p7krUF4M/repo/modules/columnflow/columnflow/util.py", line 38, in <module>
    from columnflow.types import Callable, Any, Sequence, Union, ModuleType
  File "/var/lib/condor/execute/dir_56729/job_ZhF6p7krUF4M/repo/modules/columnflow/columnflow/types.py", line 21, in <module>
    from typing_extensions import Annotated, _AnnotatedAlias as AnnotatedType  # noqa
ModuleNotFoundError: No module named 'typing_extensions'
dependency tree for branch 13 failed (exit code 1), stop job
```

Running the job with the `local` workflow works fine, which is why I think the software stack in principle is setup correctly and the `typing_extensions` module should be available. Am I doing anything wrong?

Thanks for your help!

---------------------------------------------------------------------------------------------
Setup

columnflow commit hash: 214cd0cd58da2b7ad0bbfac7a6c50f8054aa8437

I also ran `cf_update_submodules` to verify that all submodules are up to date

Platform: DESY NAF system

**Comment by pkausw** [comment]
Thanks for pointing us to this issue @monirehyamrali! I think this might be related to the centralization of the types that we recently introduced, maybe @riga can say more.

**Comment by riga** [comment]
Hi Monireh,

I'm going to debug this, the error should indeed not show up.

In the meantime, can you try adding `--htcondor-share-software` to your command? It forces the jobs to use your local software stack which should function correctly.

**Comment by riga** [comment, marked as answer]
@monirehyamrali I just fixed this. Can you

- pull the latest columnflow master into your analysis repository,
- open a new shell and before sourcing the setup, run `export CF_SANDBOX_SETUP_MODE=update` first?

This should solve it.
